rflx/pyrflx/typevalue.py

```diff
@@ -492,42 +504,69 @@ class MessageValue(TypeValue):
 
     _type: Message
 
-    def __init__(self, model: Message, refinements: Sequence[Refinement] = None) -> None:
+    def __init__(
+        self,
+        model: Message,
+        refinements: Sequence[Refinement] = None,
+        skip_verification: bool = False,
+        fields: Mapping[str, "MessageValue.Field"] = None,
+        checksums: Mapping[str, "MessageValue.Checksum"] = None,
+        type_literals: Mapping[Name, Expr] = None,
+    ) -> None:
+        # pylint: disable=too-many-arguments
         super().__init__(model)
+        self._skip_verification = skip_verification
         self._refinements = refinements or []
-        self._fields: Dict[str, MessageValue.Field] = {
+
+        self._fields: Mapping[str, MessageValue.Field] = fields if fields else {
             f.name: self.Field(
                 TypeValue.construct(
-                    self._type.types[f], imported=self._type.types[f].package != model.package
+                    self._type.types[f] if f in self._type.types else Opaque(),
+                    imported=f in self._type.types and self._type.types[f].package != model.package,
                 ),
                 f.name,
             )
-            for f in self._type.fields
+            for f in (INITIAL,) + self._type.fields
         }
 
-        self._checksums: Mapping[str, MessageValue.Checksum] = {
+        self._checksums: Mapping[str, MessageValue.Checksum] = checksums if checksums else {
             str(field_name): MessageValue.Checksum(str(field_name), parameters)
             for field_name, parameters in self._type.checksums.items()
         }
 
-        self.__type_literals: Mapping[Name, Expr] = {}
-        self._last_field: str = self._next_field(INITIAL.name)
-        for t in [
-            f.typeval.literals for f in self._fields.values() if isinstance(f.typeval, EnumValue)
-        ]:
-            self.__type_literals.update(t)
-        self.__type_literals_keys = set(self.__type_literals)
-        initial = self.Field(OpaqueValue(Opaque()), INITIAL.name)
+        self.__type_literals: Mapping[Name, Expr] = type_literals if type_literals else {}
+        if not self.__type_literals:
+            assert isinstance(self.__type_literals, dict)
+            for t in [
+                f.typeval.literals
+                for f in self._fields.values()
+                if isinstance(f.typeval, EnumValue)
+            ]:
+                self.__type_literals.update(t)
+
+        initial = self._fields[INITIAL.name]
         initial.first = Number(0)
         initial.typeval.assign(bytes())
-        self._fields[INITIAL.name] = initial
-        self._simplified_mapping: Mapping[Name, Expr] = {}
-        self._simplified_mapping_keys: Set[Name] = set()
```

Comment:
Is using a separate `self._simplified_mapping_keys` slower now?

Reply:
Since the checks that used the keys are executed much less now the overhead of adding elements to the set is higher than the gain from looking them up.